### DISCUSSION TITLE
Backend conference attendees

### DISCUSSION THREAD

**Discussion by snicki13** [thm-mni-ii/digital-classroom#61] [General]
To keep conference attendences consistent in case of a user closing the digital classroom tab, two scenarios come to mind:

1. The closing of the classroom-tab is immediately noted in the backend. It is possible to immediately remove the user from the datastructures holding information about the attending users. When the user reopens the tab it is not possible to reenter the rooms, since information about open tabs is not obtainable.
Pro: Easy to implement. Light on traffic.
Con: Not 100% accurat. User in conference is not visible to others on refresh, unless he rejoins the BBB conferences via new classroom tab. Better than leaving the users within the backends attendence list.

2. When a user connects / reconnects to the backend, the attendence of the conferences known to the backend is requested from the BBB server.
Pro: 100% Accurat.
Con: Possibly heavy traffic to BBB server & longer response times on connection. A little harder to implement.

Any thoughts on this?

**Comment by Zitrone44** [comment]
I would like to add a third alternative:
3. Use [BBB Webhook](https://docs.bigbluebutton.org/dev/webhooks.html) to get notified by the BBB-Server, when a user joins or leaves.
Pro: 100% Accurat and Light on traffic
Con: The BBB-Server needs to have `bbb-webhooks` installed

**Comment by snicki13** [comment]
Unfortunately the PILOS BBB-Backend does not support the webhook API.
I will try to support it as an alternative.
Thanks

**Comment by mxsph** [comment]
> I will try to support it as an alternative.

Yes, that would be my suggestion as well. We use option 3 if possible and option 1 otherwise. 

I find option 1 better than option 2 because the advantages that option 2 offers (100% Accurat) do not justify the disadvantages of this option.

**Comment by snicki13** [comment]
PILOS contacts the BBB backend every minute. @frankkammer suggested to do that as well.
This could be combined with option 1:
We delete a leaving user from all conferences within the classroom and readd him if the BBB backend says, that he is still attending a conference. This is possible, since there is no strong consistency enforced and a user is only identified via its unique id, which is identical in gateway, classroom and BBB backend.
